### PR TITLE
Introduce DEAL_II_NOT_IMPLEMENTED() and associated machinery.

### PR DESCRIPTION
**Pull request by bangerth**
This addresses a part of #16536: It introduces a function-like macro `DEAL_II_NOT_IMPLEMENTED()` that we can use instead of the awkward `Assert(false, ExcNotImplemented());`. The outcome is the same, but we don't have to use the fake condition.

There are changes using this macro introduces. First, the error text is different/better for not printing extraneous stuff. As an example, this is what the existing mechanism looks like:
```
--------------------------------------------------------
An error occurred in line <48> of file </home/fac/g/bangerth/p/deal.II/1/install/examples/step-1/step-1.cc> in function
    void first_grid()
The violated condition was: 
    false
Additional information: 
    You are trying to use functionality in deal.II that is currently not
    implemented. In many cases, this indicates that there simply didn't
    appear much of a need for it, or that the author of the original code
    did not have the time to implement a particular case. If you hit this
    exception, it is therefore worth the time to look into the code to
    find out whether you may be able to implement the missing
    functionality. If you do, please consider providing a patch to the
    deal.II development sources (see the deal.II website on how to
    contribute).

Stacktrace:
-----------
#0  ./step-1: first_grid()
#1  ./step-1: main
--------------------------------------------------------
```
whereas we now get this:
```
--------------------------------------------------------
An error occurred in line <48> of file </home/fac/g/bangerth/p/deal.II/1/install/examples/step-1/step-1.cc> in function
    void first_grid()
Additional information: 
    You are trying to use functionality in deal.II that is currently not
    implemented. In many cases, this indicates that there simply didn't
    appear much of a need for it, or that the author of the original code
    did not have the time to implement a particular case. If you hit this
    exception, it is therefore worth the time to look into the code to
    find out whether you may be able to implement the missing
    functionality. If you do, please consider providing a patch to the
    deal.II development sources (see the deal.II website on how to
    contribute).

Stacktrace:
-----------
#0  ./step-1: first_grid()
#1  ./step-1: main
--------------------------------------------------------
```

Second, the new macro unconditionally raises the error (i.e., either aborts the program, or throws an exception), whereas the current system lets the code run in release mode. I think this is the right decision; because there is no condition that needs to be evaluated when hitting this macro, there is also no runtime cost implied.

In cases where the question of whether something is implemented or not depends on a condition, we will continue to use
```
  Assert (cond, ExcNotImplemented());
```

This patch converts all places I could find in the library itself, but does not touch the tutorials. I'll get to that in a later PR.

### CODE REVIEW

**Comment by tamiko** [comment]
Mac OS-X is combative today...
```
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/qv/pdh5wsgn0lq3dp77zj602b5c0000gn/T/fe_bdm-3fc4a6.cpp
clang: note: diagnostic msg: /var/folders/qv/pdh5wsgn0lq3dp77zj602b5c0000gn/T/fe_bdm-3fc4a6.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /Users/runner/Library/Logs/DiagnosticReports/clang_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg: 

********************
make[2]: *** [source/fe/CMakeFiles/object_fe_debug.dir/fe_bdm.cc.o] Error 254
make[1]: *** [source/fe/CMakeFiles/object_fe_debug.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by bangerth** [comment]
Yes, I already restarted the OS X tester once. It looked like the compile job had simply been killed -- it had not complained about a specific file. Not sure what to do about it...

**Comment by bangerth** [comment]
Separate question: How do people feel about the name? I picked upper case letters both because it's a macro and because I feel that it should stand out in code. But it's negotiable.

**Comment by tamiko** [comment]
@bangerth I like `DEAL_II_NOT_IMPLEMENTED`. It would be my second choice right after `DEAL_II_HERE_BE_DRAGONS`. Honestly, a lot of things are better than `Assert(false, ...)`.

**Review by tamiko — APPROVED** [review]

File: include/deal.II/base/geometry_info.h [hunk 1]
```diff
@@ -4977,7 +4977,7 @@ inline Tensor<1, dim>
 GeometryInfo<dim>::d_linear_shape_function_gradient(const Point<dim> &,
                                                     const unsigned int)
 {
-  Assert(false, ExcNotImplemented());
+  DEAL_II_NOT_IMPLEMENTED()
```
Comment (tamiko) [inline include/deal.II/base/geometry_info.h:4980]: What about we change the syntax to `DEAL_II_NOT_IMPLEMENTED();` (with a semicolon)?

File: include/deal.II/base/exceptions.h [hunk 2]
```diff
@@ -1762,6 +1762,33 @@ namespace deal_II_exceptions
...
+#define DEAL_II_NOT_IMPLEMENTED()                                \
+  ::dealii::deal_II_exceptions::internals::issue_error_noreturn( \
+    ::dealii::deal_II_exceptions::internals::ExceptionHandling:: \
+      abort_or_throw_on_exception,                               \
+    __FILE__,                                                    \
+    __LINE__,                                                    \
+    __PRETTY_FUNCTION__,                                         \
+    nullptr,                                                     \
+    nullptr,                                                     \
+    ExcNotImplemented());
```
Comment (tamiko) [inline include/deal.II/base/exceptions.h:1789]: By removing the semicolon here.
We changed this to most of our macros recently and I very much prefer to have proper C++ syntax.

**Review by tamiko — APPROVED** [review]

**Comment by bangerth** [comment]
That is what I had meant to do anyway, and then apparently got wrong in such a way that no tester was unhappy about it.

File: include/deal.II/base/exceptions.h [hunk 3]
```diff
@@ -1762,6 +1762,33 @@ namespace deal_II_exceptions
...
+  ::dealii::deal_II_exceptions::internals::issue_error_noreturn( \
+    ::dealii::deal_II_exceptions::internals::ExceptionHandling:: \
+      abort_or_throw_on_exception,                               \
+    __FILE__,                                                    \
+    __LINE__,                                                    \
+    __PRETTY_FUNCTION__,                                         \
+    nullptr,                                                     \
+    nullptr,                                                     \
+    ExcNotImplemented())
+
```
Comment (masterleinad) [inline include/deal.II/base/exceptions.h:1790]: Does this work on GPUs? We should probably call `Kokkos::abort(error_message);` if we are on a device similar to what we are doing `Assert` already.

Comment (bangerth) [reply include/deal.II/base/exceptions.h:1790]: I can make that change -- would you be willing to let me do this in a follow-up?

Out of curiosity, what happens if you call `std::abort()` on a GPU?

Comment (masterleinad) [reply include/deal.II/base/exceptions.h:1790]: > Out of curiosity, what happens if you call std::abort() on a GPU?

It wouldn't compile, see https://godbolt.org/z/T87Pes3ox.

Comment (masterleinad) [reply include/deal.II/base/exceptions.h:1790]: > I can make that change -- would you be willing to let me do this in a follow-up?

Sure, it seems we don't have any problematic use cases in our CI yet.

**Pull request merged**